{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try Out pandas: Examine A GTF File"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's Get Situated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the GTF file into a pandas DataFrame. We've cut down this GTF file so that it only contains features found on chromosome 12."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instead of creating a new DataFrame from scratch,\n",
    "# we're going to read in the actual GTF file\n",
    "chr12_gtf = pd.read_csv('genes_ucsc.chr12.mod.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Oh man, that first attempt looked bad (why?). Let's try again\n",
    "chr12_gtf = pd.read_csv('genes_ucsc.chr12.mod.gtf', sep=\"\\t\", names=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 1: Get a feel for the DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Look at the top 4 lines of the GTF file.\n",
    "- Get the dimensions (how many rows and columns) of the GTF file.\n",
    "- Find the `dtypes` of the columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         seqname   source feature  start    end score strand frame  \\\n",
      "0  chr12_partial  unknown    exon      1    892     .      +     0   \n",
      "1  chr12_partial  unknown     CDS    162    892     .      +     0   \n",
      "2          chr12  unknown    exon  43757  43793     .      +     0   \n",
      "3          chr12  unknown     CDS  43757  43793     .      +     0   \n",
      "\n",
      "                                          attributes  \n",
      "0  gene_id \"ATXN2_partial\"; gene_name \"ATXN2_part...  \n",
      "1  gene_id \"ATXN2_partial\"; gene_name \"ATXN2_part...  \n",
      "2  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
      "3  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
      "(47113, 9)\n",
      "seqname       object\n",
      "source        object\n",
      "feature       object\n",
      "start          int64\n",
      "end            int64\n",
      "score         object\n",
      "strand        object\n",
      "frame         object\n",
      "attributes    object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "# print the top 4 lines of the GTF file\n",
    "print(chr12_gtf.head(4))\n",
    "\n",
    "# print the dimensions (row and columns) of the file\n",
    "print(chr12_gtf.shape)\n",
    "\n",
    "# print the datatypes (dtypes) of the columns\n",
    "print(chr12_gtf.dtypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2: Extract some biological information from the DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Look at the top 10 entries in the feature column of the GTF file.\n",
    "- How many entries are there for each kind of \"feature\" in this GTF file?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    exon\n",
      "1     CDS\n",
      "2    exon\n",
      "3     CDS\n",
      "4    exon\n",
      "5     CDS\n",
      "6    exon\n",
      "7     CDS\n",
      "8    exon\n",
      "9     CDS\n",
      "Name: feature, dtype: object\n",
      "exon           23203\n",
      "CDS            20284\n",
      "start_codon     1818\n",
      "stop_codon      1808\n",
      "Name: feature, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# print the top 10 entries of the feature column\n",
    "print(chr12_gtf[\"feature\"].head(10))\n",
    "\n",
    "# print the number of entries there are for each feature column\n",
    "print(chr12_gtf[\"feature\"].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3: Let's think about DataFrames as a data structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- What are some types of data you would store in a DataFrame?\n",
    "- What are some types of data you wouldn't store in a DataFrame?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### What are some types of data you would store in a DataFrame?\n",
    "The data types I would store in a DataFrame would be an indexing column that would distinguish the rest of the data (for example, a name). Other types of data I would store would be \n",
    "\n",
    "###### What are some types of data you wouldn't store in a DataFrame?\n",
    "Data I wouldn't store in a DataFrame would be extraneous data that doesn't correlate to the dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 4: Perform operations on the DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Find the length of each feature in the gtf file by i. subtracting end coordinate from start coordinate.\n",
    "- Time this operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0         891\n",
      "1         730\n",
      "2          36\n",
      "3          36\n",
      "4          59\n",
      "5          59\n",
      "6          71\n",
      "7          71\n",
      "8         150\n",
      "9         150\n",
      "10        124\n",
      "11        124\n",
      "12         91\n",
      "13         91\n",
      "14        197\n",
      "15        197\n",
      "16        178\n",
      "17        178\n",
      "18        209\n",
      "19        209\n",
      "20        182\n",
      "21        182\n",
      "22        197\n",
      "23        197\n",
      "24        107\n",
      "25        107\n",
      "26         70\n",
      "27         70\n",
      "28        304\n",
      "29        304\n",
      "         ... \n",
      "47083    4967\n",
      "47084    4967\n",
      "47085       2\n",
      "47086       2\n",
      "47087       2\n",
      "47088       2\n",
      "47089       2\n",
      "47090       2\n",
      "47091       2\n",
      "47092       2\n",
      "47093       2\n",
      "47094     659\n",
      "47095       2\n",
      "47096      83\n",
      "47097      84\n",
      "47098      84\n",
      "47099     128\n",
      "47100     128\n",
      "47101     120\n",
      "47102     120\n",
      "47103     216\n",
      "47104     216\n",
      "47105     123\n",
      "47106     123\n",
      "47107     127\n",
      "47108     127\n",
      "47109     248\n",
      "47110     270\n",
      "47111       2\n",
      "47112      29\n",
      "Length: 47113, dtype: int64\n",
      "Total time was: 0.05898165702819824\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start = time.time()\n",
    "\n",
    "# find the length of each feature in the gtf file by subtracting end coordinate from start coordinate\n",
    "print(chr12_gtf[\"end\"] - chr12_gtf[\"start\"])\n",
    "\n",
    "end = time.time()\n",
    "print(\"Total time was: \" + str(end - start))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 5: Get information about exons from the DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- a. Get all the exons in the gtf file.\n",
    "- b. For the exons, pull out the following columns:  seqname, start, end and attributes.\n",
    "- c. Try adding the .values attribute on the output from (b). It looks different from the (a) output. Why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nThe reason the .values attribute makes the specific_exon_columns non-tabular and an Array\\nis because it returns an Array version of the Pandas DataFrame object.\\n'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all exons\n",
    "exons = chr12_gtf[\"feature\"] == \"exon\"\n",
    "exons_gtf = chr12_gtf[exons]\n",
    "\n",
    "# pull out the seqname, start, end, and attributes columns\n",
    "specific_exon_columns = exons_gtf[[\"seqname\", \"start\", \"end\", \"attributes\"]]\n",
    "\n",
    "# output the specific exon columns as an Array\n",
    "specific_exon_columns.values\n",
    "\n",
    "\"\"\"\n",
    "The reason the .values attribute makes the specific_exon_columns non-tabular and an Array\n",
    "is because it returns an Array version of the Pandas DataFrame object.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 6: Get information about the gene ATXN2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Find all the rows where the `attributes` column contains `ATXN2` as a gene name\n",
    "- Narrow that down to rows where the `feature` is also a `CDS`\n",
    "- Sort these records by start coordinate\n",
    "\n",
    "**Note:** This is a little more involved. You should use Google/DuckDuckGo/ESP to search for hints on how to do this. Searching this stuff online is how we learn to do these things."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12_partial</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>162</td>\n",
       "      <td>892</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2_partial\"; gene_name \"ATXN2_part...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>43757</td>\n",
       "      <td>43793</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>45459</td>\n",
       "      <td>45518</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46699</td>\n",
       "      <td>46770</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47246</td>\n",
       "      <td>47396</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>74360</td>\n",
       "      <td>74484</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>78703</td>\n",
       "      <td>78794</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>79600</td>\n",
       "      <td>79797</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>81249</td>\n",
       "      <td>81427</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>83313</td>\n",
       "      <td>83522</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>86137</td>\n",
       "      <td>86319</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>89094</td>\n",
       "      <td>89291</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>89678</td>\n",
       "      <td>89785</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>90057</td>\n",
       "      <td>90127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>110896</td>\n",
       "      <td>111200</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>112852</td>\n",
       "      <td>112915</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>113811</td>\n",
       "      <td>113963</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>114345</td>\n",
       "      <td>114405</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>128934</td>\n",
       "      <td>129118</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>129436</td>\n",
       "      <td>129568</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>134961</td>\n",
       "      <td>135014</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>142317</td>\n",
       "      <td>142462</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>143420</td>\n",
       "      <td>143647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>145831</td>\n",
       "      <td>145999</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr12</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>146836</td>\n",
       "      <td>146864</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seqname   source feature   start     end score strand frame  \\\n",
       "1   chr12_partial  unknown     CDS     162     892     .      +     0   \n",
       "3           chr12  unknown     CDS   43757   43793     .      +     0   \n",
       "5           chr12  unknown     CDS   45459   45518     .      +     0   \n",
       "7           chr12  unknown     CDS   46699   46770     .      +     0   \n",
       "9           chr12  unknown     CDS   47246   47396     .      +     0   \n",
       "11          chr12  unknown     CDS   74360   74484     .      +     0   \n",
       "13          chr12  unknown     CDS   78703   78794     .      +     0   \n",
       "15          chr12  unknown     CDS   79600   79797     .      +     0   \n",
       "17          chr12  unknown     CDS   81249   81427     .      +     0   \n",
       "19          chr12  unknown     CDS   83313   83522     .      +     0   \n",
       "21          chr12  unknown     CDS   86137   86319     .      +     0   \n",
       "23          chr12  unknown     CDS   89094   89291     .      +     0   \n",
       "25          chr12  unknown     CDS   89678   89785     .      +     0   \n",
       "27          chr12  unknown     CDS   90057   90127     .      +     0   \n",
       "29          chr12  unknown     CDS  110896  111200     .      +     0   \n",
       "31          chr12  unknown     CDS  112852  112915     .      +     0   \n",
       "33          chr12  unknown     CDS  113811  113963     .      +     0   \n",
       "35          chr12  unknown     CDS  114345  114405     .      +     0   \n",
       "37          chr12  unknown     CDS  128934  129118     .      +     0   \n",
       "39          chr12  unknown     CDS  129436  129568     .      +     0   \n",
       "41          chr12  unknown     CDS  134961  135014     .      +     0   \n",
       "43          chr12  unknown     CDS  142317  142462     .      +     0   \n",
       "45          chr12  unknown     CDS  143420  143647     .      +     0   \n",
       "47          chr12  unknown     CDS  145831  145999     .      +     0   \n",
       "48          chr12  unknown     CDS  146836  146864     .      +     0   \n",
       "\n",
       "                                           attributes  \n",
       "1   gene_id \"ATXN2_partial\"; gene_name \"ATXN2_part...  \n",
       "3   gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "5   gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "7   gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "9   gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "11  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "13  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "15  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "17  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "19  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "21  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "23  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "25  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "27  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "29  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "31  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "33  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "35  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "37  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "39  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "41  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "43  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "45  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "47  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  \n",
       "48  gene_id \"ATXN2\"; gene_name \"ATXN2\"; p_id \"P137...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# grab only ATXN2 and CDS data-sets\n",
    "atxn2 = (chr12_gtf[\"attributes\"].str.lower().str.contains(\"atxn2\")) & (chr12_gtf[\"feature\"] == \"CDS\")\n",
    "\n",
    "# sort the atxn2 gene records by start coordinate\n",
    "atxn2_gtf = chr12_gtf[atxn2]\n",
    "atxn2_gtf = atxn2_gtf.sort_values(by=[\"start\"])\n",
    "\n",
    "# return the atxn2 gtf DataFrame\n",
    "atxn2_gtf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
